### DISCUSSION TITLE
Do you write tests for your bots? Then how?

### DISCUSSION THREAD

**Discussion by pavel-mironchik** [irazasyed/telegram-bot-sdk#1064] [Q&A] [answered]
Hello. I'm wondering if you write tests for your bots. For example, to be sure that a command works as expected. I've tried but without a success. There are 2 things I've stumbled upon:
1. Standard Laravel's test suites that use the `->post()` method don't provide data in the `php://input` to instantiate an `Update` object:
```
// vendor/irazasyed/telegram-bot-sdk/src/Methods/Update.php

private function getRequestBody(?RequestInterface $request): mixed
{
    $rawBody = $request instanceof RequestInterface ? (string) $request->getBody() : file_get_contents('php://input');

    return json_decode($rawBody, true);
}
```
2. A bot generally answers to a command by sending a request to the Telegram server. And it will send dumb requests while testing. I didn't find a way to prevent this (can we mock this?).

Will appreciate your tests examples.

I use a typical approach to handle commands:
```
// app/Http/Controllers/WebhooksHandler.php

class WebhooksHandler extends Controller
{
    public function __invoke(): string
    {
        $update = Telegram::commandsHandler(true);
        return 'ok';
    }
}
```
```
// app/TelegramBot/Commands/StartCommand.php

class StartCommand extends Command
{
    protected string $name = 'start';

    protected string $description = 'Start command';

    public function handle(): void
    {
        $this->replyWithMessage([
            'text' => 'Hello',
        ]);
    }
}
```

P.S. related issues without answers:
- https://github.com/irazasyed/telegram-bot-sdk/issues/776
- https://github.com/irazasyed/telegram-bot-sdk/issues/201

@irazasyed @jonnywilliamson @lptn @ihoru ?

**Comment by jonnywilliamson** [comment, marked as answer]
In case this is some help for you.

I'm still at the begining of my testing learning career, and starting with PEST so what I'm doing might not be proper. But in general, I've found I can test what gets sent to my bot for delivery to a user. 

Here's an example of one of my tests at the moment.

```php
it('should notify admin if retrieving a flight duty statistics fails', function () {
    $user = User::factory()->create(['telegram_id' => 987654321]);
    $botMock = Mockery::mock(Bot::class)->expects('sendMessage')->with(Mockery::capture($params))->getMock();
    Telegram::shouldReceive(['bot' => $botMock]);

    //Doing some action that will send a message or photo or whatever at some stage during the process that I want to check
    //has the right info.
    $results = app(FlightTimeStatisticsAction::class)->for($user);

    expect($results)->toHaveCount(0);
    expect($results)->toBeInstanceOf(Collection::class);
    expect($params['chat_id'])->toBe(987654321);
    expect($params['text'])->toContain('Unable to parse flight/duty time statistics');
});
```

Also bare in mind that I'm using the newer v4 (not fully released) version of the sdk and don't use any previous versions any more.
